### PR TITLE
Added semantically equivalent comparison to SemanticallyEqual.

### PR DESCRIPTION
**Pull request by traceyyoshima**
Changes:

- Added semantically equivalent comparison to `SemanticallyEqual`.
- Added clean up recipe `CombineSemanticallyEqualCatchBlocks`.

### CODE REVIEW

File: rewrite-test/src/main/kotlin/org/openrewrite/java/cleanup/CombineSemanticallyEqualCatchBlocksTest.kt [hunk 1]
```diff
@@ -0,0 +1,264 @@
...
+                        throw new RuntimeException(ex);
+                    } catch (RuntimeException ex) {
+                        checkEquivalence("diff");
+                        throw new RuntimeException(ex);
+                    }
+                }
+                void checkEquivalence(String s) {
+                }
+            }
+        """)
```
Comment (sambsnyd) [inline rewrite-test/src/main/kotlin/org/openrewrite/java/cleanup/CombineSemanticallyEqualCatchBlocksTest.kt:116]: Changing the order the catch blocks are evaluated in isn't always safe and can represent a change in program semantics. Catch blocks are evaluated in order and the first matching block is taken. 

Given exception types `A`, `B`, `C` caught in that order, it is safe to transform into `A | C`, `B` _only_ if `C` is not a subtype/implementation of `B` and therefore could never be handled by that catch.

**Review by sambsnyd — CHANGES_REQUESTED** [review]

File: rewrite-test/src/main/kotlin/org/openrewrite/java/cleanup/CombineSemanticallyEqualCatchBlocksTest.kt (hunk 1 above)
Comment (traceyyoshima) [reply rewrite-test/src/main/kotlin/org/openrewrite/java/cleanup/CombineSemanticallyEqualCatchBlocksTest.kt:116]: That's fair, I'll update the qualification combine catch blocks to consider the previous caught types.

**Comment by traceyyoshima** [comment]
Update:
I updated the PR with feedback from the team.

The recipe requires changes from https://github.com/openrewrite/rewrite/issues/1896 and to use `maybeAutoformat`.
Lastly, no changes should be made if comments cause the blocks to be different.

I plan to switch gears to focus on higher priority issues and will come back to this afterward.

File: rewrite-java/src/main/java/org/openrewrite/java/cleanup/CombineSemanticallyEqualCatchBlocks.java [hunk 2]
```diff
@@ -0,0 +1,323 @@
...
+    }
+
+    @Override
+    protected TreeVisitor<?, ExecutionContext> getVisitor() {
+        return new CombineSemanticallyEqualCatchBlocksVisitor();
+    }
+
+    private static class CombineSemanticallyEqualCatchBlocksVisitor extends JavaVisitor<ExecutionContext> {
+
+        // TODO: account for comments.
```
Comment (traceyyoshima) [inline rewrite-java/src/main/java/org/openrewrite/java/cleanup/CombineSemanticallyEqualCatchBlocks.java:60]: @sambsnyd, you mentioned `SemanticallyEqual` should not contain a mechanism to consider comments while comparing equality or equivalency.

Without the changes, I need to copy `SemanticallyEqual` into `CombineSemanticallyEqualCatchBlocksVisitor` and make a few adjustments to compare comments between the J trees.

Do you think it's okay to copy the entire class into this recipe until we see a need to account for comments in `SemanticallyEqual`?

**Pull request merged**